File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    # 위 아래 횟수
    manual = {
        "A": 0,
        "B": 1,
        "C": 2,
        "D": 3,
        "E": 4,
        "F": 5,
        "G": 6,
        "H": 7,
        "I": 8,
        "J": 9,
        "K": 10,
        "L": 11,
        "M": 12,
        "N": 13,
        "O": 12,
        "P": 11,
        "Q": 10,
        "R": 9,
        "S": 8,
        "T": 7,
        "U": 6,
        "V": 5,
        "W": 4,
        "X": 3,
        "Y": 2,
        "Z": 1
    }
    
    # 위 아래 조작은 무관
    cnt = 0
    for c in name:
        cnt += manual[c]
    
    # 모두 오른쪽으로 움직인 경우
    min_moves = len(name) -1
    
    for idx in range(len(name)):
        # idx : 움직인 횟수
        next_idx = idx +1
        # 다음 A값 구하기
        while next_idx < len(name) and name[next_idx] == "A":
            next_idx +=1
            
        min_moves = min(min_moves, 2*idx + len(name) - next_idx, idx + 2*(len(name) - next_idx))
    
    
    return cnt + min_moves

File: 프로그래머스/2/42860. 조이스틱/조이스틱.js
// 좌 우 판단 기준?
// 위 아래 판단 기준 -> 알파벳 m보다 작은 경우
const calUpDown = (target) => {
    if(target.charCodeAt(0) >= 'N'.charCodeAt(0)){
        return 'Z'.charCodeAt(0) - target.charCodeAt(0) + 1;
    }
    return target.charCodeAt(0) - 'A'.charCodeAt(0);
}

function solution(name) {
    let upDownCount = 0;
    for(let i = 0; i < name.length; i++){
        upDownCount += calUpDown(name[i]);
    }
    
    // 오른쪽으로 쭉 이동하는 경우에서 뒤로 돌아오는 경우가 더 적은지를 체크한다.
     let minMoveCount = name.length - 1; // 오른쪽으로만 쭉 가는 경우
    
    for (let i = 0; i < name.length; i++) {
        // 'A'가 아닌 첫 번째 문자의 위치
        let next = i + 1;
        while (next < name.length && name[next] === 'A') {
            next++;
        }
        
        // 오른쪽으로 갔다가 돌아오는 경우
        const move1 = (i * 2) + (name.length - next);
        
        // 왼쪽으로 갔다가 돌아오는 경우
        const move2 = (name.length - next) * 2 + i;
        
        minMoveCount = Math.min(minMoveCount, move1, move2);
    }
    
    return upDownCount + minMoveCount;
}